### ISSUE TITLE
Sharding Datasets Outside of TFDS

### ISSUE THREAD

**Issue by carlesoctav** [apple/axlearn#1329]
Hi, I have a question regarding the dataset processing workflow.

Currently, I know that in this codebase we shard by using TFDS shard options directly on single data files (potentially very large datasets). From what I understand, each process loads a different batch. At the end of the process, we need to use jax.make_array_from_process_local_data to combine the batches into a single global batch with global sharding (across multiple devices in multiple processes), especially if the model is also sharded across global devices .

One thing I dislike about this approach is that we download the full dataset for each process (correct me if I'm wrong).
What do you think about this alternative workflow: we shard the data file before passing it to TFDS. For each process (probably just the processes/machines related to DP, FSDP on the process axis), we load only the assigned dataset portion (for example, load ds_1 for process 1, etc.). Then we create a global batch with make_global_array.

One problem I've encountered with this setup is that each "same variable" on different processes should have the same array. I'm not sure how to bypass this issue (is it because of the multiprocess/multihost architecture of jax?).

Do you have any experience with this approach?
The problem I want to solve: I have TPUs, but each host only has ~100GB disk space while having 300+ GB RAM. My current setup involves creating an NFS host in the same region and attaching this NFS to each host.
